app/src/main/java/dominando/android/livros/BookFormFragment.kt

```diff
@@ -33,78 +31,53 @@ class BookFormFragment : BaseFragment() {
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
-    ): View? {
-        binding = DataBindingUtil.inflate(
-                inflater, R.layout.fragment_book_form, container, false
-        )
-        return binding.root
+    ): View {
+        val binding = DataBindingUtil.inflate(
+                inflater,
+                R.layout.fragment_book_form,
+                container, false) as FragmentBookFormBinding
+
+        return binding.run {
+            lifecycleOwner = this@BookFormFragment
+            content.presenter = this@BookFormFragment
+            content.viewModel = this@BookFormFragment.viewModel
+            root
+        }
     }
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
-        if (viewModel.book == null) {
-            viewModel.book = arguments?.getParcelable("book") ?: Book()
+        arguments?.getParcelable<Book>("book")?.let {
+            viewModel.setBook(it)
         }
-        binding.content.book = viewModel.book
-        binding.content.publishers = listOf(
-                Publisher("1", "Novatec"),
-                Publisher("2", "Outra")
-        )
-        binding.content.presenter = this
-        binding.lifecycleOwner = this
         lifecycle.addObserver(viewModel)
         init()
     }
 
     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
         super.onActivityResult(requestCode, resultCode, data)
         if (resultCode == Activity.RESULT_OK && requestCode == RC_CAMERA) {
-            binding.content.book?.coverUrl = "file://${viewModel.tempImageFile?.absolutePath}"
+            viewModel.book().value?.coverUrl = "file://${viewModel.tempImageFile?.absolutePath}"
         }
     }
 
     private fun init() {
-        viewModel.getState().observe(viewLifecycleOwner, Observer { event ->
-            event?.peekContent()?.let { state ->
+        viewModel.state.observe(viewLifecycleOwner, Observer { event ->
```

Comment:
Running the app, I noticed that a message is displayed if the user try to save a book with empty fields.
If the device is rotated, the message is displayed again.

Reply:
I have seen this bug.
To solve it I needed to create the `SingleLiveEvent` class. This is Google's solution to the one-shot event using LiveData. 
https://github.com/android/architecture-samples/blob/dev-todo-mvvm-live/todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/SingleLiveEvent.java

So we don't need to use `LiveEvent` class anymore